packages/override-tools/src/Manifest.ts

```diff
@@ -0,0 +1,161 @@
+ */
+export async function readFromFile(filePath: string): Promise<Manifest> {
+  const json = (await fs.readFile(filePath)).toString();
+  return this.parse(json);
+}
+
+/**
+ * Parse a string with JSON for the override manifest into one.
+ *
+ * @throws if the JSON doesn't describe a valid manifest
+ */
+export function parse(json: string): Manifest {
+  const parsed = JSON.parse(json);
+
+  ThrowReporter.report(ManifestType.decode(parsed));
+  return parsed;
+}
+
+export interface ValidationError {
+  type:
+    | 'fileMissingFromManifest' // An override file is present with no manifest entry
+    | 'overrideFileNotFound' // The manifest describes a file which does not exist
+    | 'baseFileNotFound' // The base file for a manifest entry cannot be found
+    | 'outOfDate'; // A base file has changed since the manifested version
+  file: string;
+}
+
+/**
+ * Check that overrides are accurately accounted for in the manifest. I.e. we
+ * should have a 1:1 mapping between files and manifest entries, and base files
+ * should be present and unchanged since entry creation.
+ */
+export async function validate(
+  manifest: Manifest,
+  overrideRepo: IOverrideFileRepository,
+  reactRepo: IReactFileRepository,
+): Promise<Array<ValidationError>> {
+  const errors: Array<ValidationError> = [];
+
+  const manifestedFiles = manifest.overrides.map(override => override.file);
+  const overrideFiles = await overrideRepo.listFiles();
+
+  const fileMissingFromManifest = _.difference(overrideFiles, manifestedFiles);
+  fileMissingFromManifest.forEach(file =>
+    errors.push({type: 'fileMissingFromManifest', file: file}),
+  );
+  const overridesNotFound = _.difference(manifestedFiles, overrideFiles);
+  overridesNotFound.forEach(file =>
+    errors.push({type: 'overrideFileNotFound', file: file}),
+  );
+
+  manifest = sortIntoVersionBatches(manifest);
+  await Promise.all(
+    manifest.overrides.map(async override => {
+      if (override.type === 'platform') {
+        return;
+      }
+
+      const baseContent = await reactRepo.getFileContents(override.baseFile);
+      if (baseContent === null) {
+        errors.push({type: 'baseFileNotFound', file: override.baseFile});
+        return;
+      }
+
+      const baseHash = hashContent(baseContent);
+      if (baseHash.toLowerCase() !== override.baseHash.toLowerCase()) {
+        errors.push({type: 'outOfDate', file: override.file});
+        return;
+      }
+    }),
+  );
+
+  return errors;
+}
+
+/**
+ * Hash content into the form expected in a manifest entry. Exposed for
+ * testing.
+ */
+export function hashContent(str: string) {
+  const hasher = crypto.createHash('sha1');
+  hasher.update(str);
+  return hasher.digest('hex');
+}
+
+/**
+ * GitReactFileRepository takes time to switch between versions. Batch versions
+ * together to reduce the number of checkouts that happen.
+ */
+function sortIntoVersionBatches(manifest: Manifest): Manifest {
```

Comment:
I should be able to remove this. We want to do that later for some things, but we don't actually check out different versions for verification.